nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -1,42 +1,36 @@
 #! @python3@/bin/python3 -B
-import argparse
-import shutil
-import os
-import sys
-import errno
-import subprocess
+
+from argparse import ArgumentParser
+from datetime import datetime
+
 import glob
-import tempfile
-import errno
-import warnings
-import ctypes
-libc = ctypes.CDLL("libc.so.6")
 import re
-import datetime
-import glob
-import os.path
-from typing import Tuple, List, Optional
 
-SystemIdentifier = Tuple[Optional[str], int, Optional[str]]
+import os
+from shutil import copyfile
+from ctypes import CDLL
 
+syncfs = CDLL("libc.so.6").syncfs
 
-def copy_if_not_exists(source: str, dest: str) -> None:
-    if not os.path.exists(dest):
-        shutil.copyfile(source, dest)
+import subprocess
 
+from sys import stderr
+from warnings import warn
 
-def generation_dir(profile: Optional[str], generation: int) -> str:
-    if profile:
-        return "/nix/var/nix/profiles/system-profiles/%s-%d-link" % (profile, generation)
-    else:
-        return "/nix/var/nix/profiles/system-%d-link" % (generation)
+from typing import NamedTuple, Optional
 
-def system_dir(profile: Optional[str], generation: int, specialisation: Optional[str]) -> str:
-    d = generation_dir(profile, generation)
-    if specialisation:
-        return os.path.join(d, "specialisation", specialisation)
-    else:
-        return d
+class SystemIdentifier(NamedTuple):
+    profile: Optional[str]
+    generation: int
+    specialisation: Optional[str]
+
+
+def profile_path(profile: Optional[str]) -> str:
+    return "/nix/var/nix/profiles/" + (f"system-profiles/{profile}" if profile else "system")
+
+def system_dir(gen: SystemIdentifier) -> str:
+    generation_dir = f"{profile_path(gen.profile)}-{gen.generation}-link"
+    return f"{generation_dir}/specialisation/{gen.specialisation}" if gen.specialisation else generation_dir
```

Comment:
For clarity I would
```suggestion
    specialisation_dir = f"{generation_dir}/specialisation/{gen.specialisation}"
    return specialisation_dir if gen.specialisation else generation_dir
```

Also, pathlib.